src/go/util/httppattern/uri_template.go

```diff
@@ -45,365 +47,100 @@ type variable struct {
 	FieldPath []string
 
 	// Do we have a ** in the variable template?
-	HasWildcardPath bool
+	HasDoubleWildCard bool
 }
 
-// HTTP Template Grammar:
-// Questions:
-//
-// Template = "/" | "/" Segments [ Verb ] ;
-// Segments = Segment { "/" Segment } ;
-// Segment  = "*" | "**" | LITERAL | Variable ;
-// Variable = "{" FieldPath [ "=" Segments ] "}" ;
-// FieldPath = IDENT { "." IDENT } ;
-// Verb     = ":" LITERAL ;
-type parser struct {
-	input      string
-	tb         int
-	te         int
-	inVariable bool
-	segments   []string
-	verb       string
-	variables  []*variable
-}
-
-func Parse(ht string) *UriTemplate {
-	if ht == "/" {
-		return &UriTemplate{}
+func ReplaceVariableFieldInUriTemplate(input string, variableFieldMapping map[string]string) (string, error) {
+	uriTemplate := ParseUriTemplate(input)
+	if uriTemplate == nil {
+		return "", fmt.Errorf("invalid uri template `%s`", input)
 	}
 
-	p := parser{
-		input: ht,
-	}
-	if !p.parse() || !p.validateParts() {
-		return nil
-	}
+	uriTemplate.replaceVariableField(variableFieldMapping)
 
-	return &UriTemplate{
-		Segments:  p.segments,
-		Verb:      p.verb,
-		Variables: p.variables,
-	}
+	return uriTemplate.String(), nil
 }
 
-func (p *parser) parse() bool {
-	if !p.parseTemplate() || !p.consumeAllInput() {
-		return false
+func (u *UriTemplate) String() string {
```

Comment:
How about ToString()?

Reply:
`String()` is the the `ToString()` in [golang](https://stackoverflow.com/questions/41232904/is-there-a-go-equivalent-of-tostring-that-fmt-print-will-use).